src/output-json-alert.c

```diff
@@ -169,18 +171,42 @@ static void AlertJsonSourceTarget(const Packet *p, const PacketAlert *pa,
     jb_close(js);
 }
 
-static void AlertJsonMetadata(AlertJsonOutputCtx *json_output_ctx,
-        const PacketAlert *pa, JsonBuilder *js)
+static void AlertJsonReference(const PacketAlert *pa, JsonBuilder *jb)
+{
+    if (!pa->s->references) {
+        return;
+    }
+
+    const DetectReference *kv = pa->s->references;
+    jb_open_array(jb, "references");
+    while (kv) {
+        /* Note that the key and reference sizes have been bound
+         * checked during parsing
+         */
+        const size_t size_needed = kv->key_len + kv->reference_len + 1;
+        char kv_store[size_needed];
+        // Some reference values may contain the scheme (unnecessarily, but
+        // allowed) so check for that here.
+        if (kv->reference_len >= kv->key_len && strncmp(kv->reference, kv->key, kv->key_len) == 0)
+            snprintf(kv_store, size_needed, "%s", kv->reference);
+        else
+            snprintf(kv_store, size_needed, "%s%s", kv->key, kv->reference);
+        jb_append_string(jb, kv_store);
```

Comment:
While trying to confirm whether one of [Jason's comments](https://github.com/OISF/suricata/pull/11638/files/6cba04a87be04e8923f121d7e23e1a51fe8cc50a#r1727814117) had been addressed, I noticed that we leave the sizes check for when the values are parsed. Is that being done [here](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/src/detect-reference.c#L121-L129)?

Reply:
What's being done here is to account for many cases where the reference information includes the key. Usually, a reference is `key,value` --> `url,www.google.com/some-link`. Url is `https://` so this reference is expanded to `https://www.google.com/some-link`

There are references that are incorrectly listed in a rules file such as `url,https://www.google.com/some-link`. The logic here detects that and avoids adding the `url` value if it's already present.